File: algorithms/2_medium/F/01673_find_the_most_competitive_subsequence/stack.cpp
class Solution {
public:
    static vector<int> mostCompetitive(const vector<int>& nums, const int k) {
        assert(k >= 1 && k <= std::ssize(nums));
        const ptrdiff_t len = std::ssize(nums);

        auto stack = vector<int>();
        stack.reserve(k);

        for (ptrdiff_t i = 0; i < len; i += 1) {
            const int num = nums[i];
            const ptrdiff_t remains = len - i;

            while (!stack.empty() && remains + std::ssize(stack) > k && stack.back() > num) {
                stack.pop_back();
            }

            if (std::ssize(stack) < k) {
                stack.push_back(num);
            }
        }

        return stack;
    }
};

File: algorithms/2_medium/F/01673_find_the_most_competitive_subsequence/stack.rs
impl Solution {
    pub fn most_competitive(nums: Vec<i32>, k: i32) -> Vec<i32> {
        assert!(k >= 1);
        let len = nums.len();
        let k = k as usize;
        assert!(k <= len);

        let mut stack: Vec<i32> = Vec::with_capacity(k);

        for (i, &num) in nums.iter().enumerate() {
            let remains = len - i;

            while !stack.is_empty() && remains + stack.len() > k && *stack.last().unwrap() > num {
                stack.pop();
            }

            if stack.len() < k {
                stack.push(num);
            }
        }

        stack
    }
}
